File: clients/android/NewsBlur/app/src/main/java/com/newsblur/util/ReadingAction.java
package com.newsblur.util;

import static com.newsblur.service.NbSyncManager.UPDATE_INTEL;
import static com.newsblur.service.NbSyncManager.UPDATE_METADATA;
import static com.newsblur.service.NbSyncManager.UPDATE_SOCIAL;
import static com.newsblur.service.NbSyncManager.UPDATE_STORY;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.Serializable;

import com.newsblur.database.BlurDatabaseHelper;
import com.newsblur.database.DatabaseConstants;
import com.newsblur.domain.Classifier;
import com.newsblur.network.domain.CommentResponse;
import com.newsblur.network.domain.NewsBlurResponse;
import com.newsblur.network.domain.StoriesResponse;
import com.newsblur.network.APIManager;
import com.newsblur.service.NbSyncManager;
import com.newsblur.service.NBSyncService;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class ReadingAction implements Serializable {

    private static final long serialVersionUID = 0L;

    private enum ActionType {
        MARK_READ,
        MARK_UNREAD,
        SAVE,
        UNSAVE,
        SHARE,
        UNSHARE,
        REPLY,
        EDIT_REPLY,
        DELETE_REPLY,
        LIKE_COMMENT,
        UNLIKE_COMMENT,
        MUTE_FEEDS,
        UNMUTE_FEEDS,
        SET_NOTIFY,
        INSTA_FETCH,
        UPDATE_INTEL,
        RENAME_FEED
    };

    private long time;
    private int tried;
    private ActionType type;
    private String storyHash;
    private FeedSet feedSet;
    private Long olderThan;
    private Long newerThan;
    private String storyId;
    private String feedId;
    private String sourceUserId;
    private String commentReplyText; // used for both comments and replies
    private String commentUserId;
    private String replyId;
    private String notifyFilter;
    private List<String> notifyTypes;
    private List<String> userTags;
    private Classifier classifier;
    private String newFeedName;

    // For mute/unmute the API call is always the active feed IDs.
    // We need the feed Ids being modified for the local call.
    private Set<String> activeFeedIds;
    private Set<String> modifiedFeedIds;

    private ReadingAction() {
        // note: private - must use helpers
        this(System.currentTimeMillis(), 0);
    }

    private ReadingAction(long time, int tried) {
        // note: private - must use helpers
        this.time = time;
        this.tried = tried;
    }
    
    public int getTried() {
        return tried;
    }

    public static ReadingAction markStoryRead(@Nullable String hash) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.MARK_READ;
        ra.storyHash = hash;
        return ra;
    }

    public static ReadingAction markStoryUnread(@Nullable String hash) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.MARK_UNREAD;
        ra.storyHash = hash;
        return ra;
    }

    public static ReadingAction saveStory(@Nullable String hash, @Nullable List<String> userTags) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.SAVE;
        ra.storyHash = hash;
        if (userTags == null) {
            ra.userTags = new ArrayList<>();
        } else {
            ra.userTags = userTags;
        }
        return ra;
    }

    public static ReadingAction unsaveStory(@Nullable String hash) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.UNSAVE;
        ra.storyHash = hash;
        return ra;
    }

    public static ReadingAction markFeedRead(@NonNull FeedSet fs, @Nullable Long olderThan, @Nullable Long newerThan) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.MARK_READ;
        ra.feedSet = fs;
        ra.olderThan = olderThan;
        ra.newerThan = newerThan;
        return ra;
    }

    public static ReadingAction shareStory(@Nullable String hash, @Nullable String storyId, @Nullable String feedId, @Nullable String sourceUserId, @Nullable String commentReplyText) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.SHARE;
        ra.storyHash = hash;
        ra.storyId = storyId;
        ra.feedId = feedId;
        ra.sourceUserId = sourceUserId;
        ra.commentReplyText = commentReplyText;
        return ra;
    }

    public static ReadingAction unshareStory(@Nullable String hash, @Nullable String storyId, @Nullable String feedId) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.UNSHARE;
        ra.storyHash = hash;
        ra.storyId = storyId;
        ra.feedId = feedId;
        return ra;
    }

    public static ReadingAction likeComment(@Nullable String storyId, @Nullable String commentUserId, @Nullable String feedId) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.LIKE_COMMENT;
        ra.storyId = storyId;
        ra.commentUserId = commentUserId;
        ra.feedId = feedId;
        return ra;
    }

    public static ReadingAction unlikeComment(@Nullable String storyId, @Nullable String commentUserId, @Nullable String feedId) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.UNLIKE_COMMENT;
        ra.storyId = storyId;
        ra.commentUserId = commentUserId;
        ra.feedId = feedId;
        return ra;
    }

    public static ReadingAction replyToComment(@Nullable String storyId, @Nullable String feedId, @Nullable String commentUserId, @Nullable String commentReplyText) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.REPLY;
        ra.storyId = storyId;
        ra.commentUserId = commentUserId;
        ra.feedId = feedId;
        ra.commentReplyText = commentReplyText;
        return ra;
    }

    public static ReadingAction updateReply(@Nullable String storyId, @Nullable String feedId, @Nullable String commentUserId, @Nullable String replyId, @Nullable String commentReplyText) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.EDIT_REPLY;
        ra.storyId = storyId;
        ra.commentUserId = commentUserId;
        ra.feedId = feedId;
        ra.commentReplyText = commentReplyText;
        ra.replyId = replyId;
        return ra;
    }

    public static ReadingAction deleteReply(@Nullable String storyId, @Nullable String feedId, @Nullable String commentUserId, @Nullable String replyId) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.DELETE_REPLY;
        ra.storyId = storyId;
        ra.commentUserId = commentUserId;
        ra.feedId = feedId;
        ra.replyId = replyId;
        return ra;
    }

    public static ReadingAction muteFeeds(@NonNull Set<String> activeFeedIds, @NonNull Set<String> modifiedFeedIds) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.MUTE_FEEDS;
        ra.activeFeedIds = activeFeedIds;
        ra.modifiedFeedIds = modifiedFeedIds;
        return ra;
    }

    public static ReadingAction unmuteFeeds(@NonNull Set<String> activeFeedIds, @NonNull Set<String> modifiedFeedIds) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.UNMUTE_FEEDS;
        ra.activeFeedIds = activeFeedIds;
        ra.modifiedFeedIds = modifiedFeedIds;
        return ra;
    }

    public static ReadingAction setNotify(@Nullable String feedId, @Nullable List<String> notifyTypes, @Nullable String notifyFilter) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.SET_NOTIFY;
        ra.feedId = feedId;
        if (notifyTypes == null) {
            ra.notifyTypes = new ArrayList<String>();
        } else {
            ra.notifyTypes = notifyTypes;
        }
        ra.notifyFilter = notifyFilter;
        return ra;
    }

    public static ReadingAction instaFetch(@Nullable String feedId) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.INSTA_FETCH;
        ra.feedId = feedId;
        return ra;
    }

    public static ReadingAction updateIntel(@Nullable String feedId, @Nullable Classifier classifier, @Nullable FeedSet fs) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.UPDATE_INTEL;
        ra.feedId = feedId;
        ra.classifier = classifier;
        ra.feedSet = fs;
        return ra;
    }

    public static ReadingAction renameFeed(@Nullable String feedId, @Nullable String newFeedName) {
        ReadingAction ra = new ReadingAction();
        ra.type = ActionType.RENAME_FEED;
        ra.feedId = feedId;
        ra.newFeedName = newFeedName;
        return ra;
    }

	public ContentValues toContentValues() {
		ContentValues values = new ContentValues();
        values.put(DatabaseConstants.ACTION_TIME, time);
        values.put(DatabaseConstants.ACTION_TRIED, tried);
        // because ReadingActions will have to represent a wide and ever-growing variety of interactions,
        // the number of parameters will continue growing unbounded.  to avoid having to frequently modify the
        // database and support a table with dozens or hundreds of columns that are only ever used at a low
        // cardinality, only the ACTION_TIME and ACTION_TRIED values are stored in columns of their own, and
        // all remaining fields are frozen as JSON, since they are never queried upon.
        values.put(DatabaseConstants.ACTION_PARAMS, DatabaseConstants.JsonHelper.toJson(this));
		return values;
	}

	public static ReadingAction fromCursor(@NonNull Cursor c) {
        long time = c.getLong(c.getColumnIndexOrThrow(DatabaseConstants.ACTION_TIME));
        int tried = c.getInt(c.getColumnIndexOrThrow(DatabaseConstants.ACTION_TRIED));
        String params = c.getString(c.getColumnIndexOrThrow(DatabaseConstants.ACTION_PARAMS));
		ReadingAction ra = DatabaseConstants.JsonHelper.fromJson(params, ReadingAction.class);
        ra.time = time;
        ra.tried = tried;
		return ra;
	}

    /**
     * Execute this action remotely via the API.
     */
    public NewsBlurResponse doRemote(@NonNull APIManager apiManager, @NonNull BlurDatabaseHelper dbHelper, @NonNull StateFilter stateFilter) {
        // generic response to return
        NewsBlurResponse result = null;
        // optional specific responses that are locally actionable
        StoriesResponse storiesResponse = null;
        CommentResponse commentResponse = null;
        int impact = 0;
        switch (type) {

            case MARK_READ:
                if (storyHash != null) {
                    result = apiManager.markStoryAsRead(storyHash);
                } else if (feedSet != null) {
                    result = apiManager.markFeedsAsRead(feedSet, olderThan, newerThan);
                }
                break;
                
            case MARK_UNREAD:
                result = apiManager.markStoryHashUnread(storyHash);
                break;

            case SAVE:
                result = apiManager.markStoryAsStarred(storyHash, userTags);
                break;

            case UNSAVE:
                result = apiManager.markStoryAsUnstarred(storyHash);
                break;

            case SHARE:
                storiesResponse = apiManager.shareStory(storyId, feedId, commentReplyText, sourceUserId);
                break;

            case UNSHARE:
                storiesResponse = apiManager.unshareStory(storyId, feedId);
                break;

            case LIKE_COMMENT:
                result = apiManager.favouriteComment(storyId, commentUserId, feedId);
                break;

            case UNLIKE_COMMENT:
                result = apiManager.unFavouriteComment(storyId, commentUserId, feedId);
                break;

            case REPLY:
                commentResponse = apiManager.replyToComment(storyId, feedId, commentUserId, commentReplyText);
                break;

            case EDIT_REPLY:
                 commentResponse = apiManager.editReply(storyId, feedId, commentUserId, replyId, commentReplyText);
                break;

            case DELETE_REPLY:
                commentResponse = apiManager.deleteReply(storyId, feedId, commentUserId, replyId);
                break;

            case MUTE_FEEDS:
            case UNMUTE_FEEDS:
                result = apiManager.saveFeedChooser(activeFeedIds);
                break;

            case SET_NOTIFY:
                result = apiManager.updateFeedNotifications(feedId, notifyTypes, notifyFilter);
                break;

            case INSTA_FETCH:
                result = apiManager.instaFetch(feedId);
                // also trigger a recount, which will unflag the feed as pending
                NBSyncService.addRecountCandidates(FeedSet.singleFeed(feedId));
                NBSyncService.flushRecounts();
                break;

            case UPDATE_INTEL:
                result = apiManager.updateFeedIntel(feedId, classifier);
                // also reset stories for the calling view so they get new scores
                NBSyncService.resetFetchState(feedSet);
                // and recount unreads to get new focus counts
                NBSyncService.addRecountCandidates(feedSet);
                break;

            case RENAME_FEED:
                result = apiManager.renameFeed(feedId, newFeedName);
                break;

            default:
                throw new IllegalStateException("cannot execute uknown type of action.");

        }
        
        if (storiesResponse != null) {
            result = storiesResponse;
            if (storiesResponse.story != null) {
                dbHelper.updateStory(storiesResponse, stateFilter, true);
            } else {
                com.newsblur.util.Log.w(this, "failed to refresh story data after action");
            }
            impact |= NbSyncManager.UPDATE_SOCIAL;
        }
        if (commentResponse != null) {
            result = commentResponse;
            if (commentResponse.comment != null) {
                dbHelper.updateComment(commentResponse, storyId);
            } else {
                com.newsblur.util.Log.w(this, "failed to refresh comment data after action");
            }
            impact |= NbSyncManager.UPDATE_SOCIAL;
        }
        if (result != null && impact != 0) {
            result.impactCode = impact;
        }
        return result;
    }

    public int doLocal(@NonNull Context context, @NonNull BlurDatabaseHelper dbHelper) {
        return doLocal(context, dbHelper, false);
    }

    /**
     * Excecute this action on the local DB. These *must* be idempotent.
     *
     * @param isFollowup flag that this is a double-check invocation and is noncritical
     *
     * @return the union of update impact flags that resulted from this action.
     */
    public int doLocal(@NonNull Context context, @NonNull BlurDatabaseHelper dbHelper, boolean isFollowup) {
        String userId = PrefsUtils.getUserId(context);
        int impact = 0;
        switch (type) {

            case MARK_READ:
                if (storyHash != null) {
                    dbHelper.setStoryReadState(storyHash, true);
                } else if (feedSet != null) {
                    dbHelper.markStoriesRead(feedSet, olderThan, newerThan);
                    dbHelper.updateLocalFeedCounts(feedSet);
                }
                impact |= UPDATE_METADATA;
                impact |= UPDATE_STORY;
                break;
                
            case MARK_UNREAD:
                dbHelper.setStoryReadState(storyHash, false);
                impact |= UPDATE_METADATA;
                break;

            case SAVE:
                dbHelper.setStoryStarred(storyHash, userTags, true);
                impact |= UPDATE_METADATA;
                break;

            case UNSAVE:
                dbHelper.setStoryStarred(storyHash, null, false);
                impact |= UPDATE_METADATA;
                break;

            case SHARE:
                if (isFollowup) break; // shares are only placeholders
                dbHelper.setStoryShared(storyHash, userId, true);
                dbHelper.insertCommentPlaceholder(storyId, userId, commentReplyText);
                impact |= UPDATE_SOCIAL;
                impact |= UPDATE_STORY;
                break;

            case UNSHARE:
                dbHelper.setStoryShared(storyHash, userId, false);
                dbHelper.clearSelfComments(storyId, userId);
                impact |= UPDATE_SOCIAL;
                impact |= UPDATE_STORY;
                break;

            case LIKE_COMMENT:
                dbHelper.setCommentLiked(storyId, commentUserId, userId, true);
                impact |= UPDATE_SOCIAL;
                break;

            case UNLIKE_COMMENT:
                dbHelper.setCommentLiked(storyId, commentUserId, userId, false);
                impact |= UPDATE_SOCIAL;
                break;

            case REPLY:
                if (isFollowup) break; // replies are only placeholders
                dbHelper.insertReplyPlaceholder(storyId, userId, commentUserId, commentReplyText);
                break;

            case EDIT_REPLY:
                dbHelper.editReply(replyId, commentReplyText);
                impact |= UPDATE_SOCIAL;
                break;

            case DELETE_REPLY:
                dbHelper.deleteReply(replyId);
                impact |= UPDATE_SOCIAL;
                break;
                
            case MUTE_FEEDS:
            case UNMUTE_FEEDS:
                dbHelper.setFeedsActive(modifiedFeedIds, type == ActionType.UNMUTE_FEEDS);
                impact |= UPDATE_METADATA;
                break;

            case SET_NOTIFY:
                impact |= UPDATE_METADATA;
                break;

            case INSTA_FETCH:
                if (isFollowup) break; // non-idempotent and purely graphical
                dbHelper.setFeedFetchPending(feedId);
                break;

            case UPDATE_INTEL:
                // TODO: because intel is always calculated on the server, we can change the disposition of
                // individual tags and authors etc in the UI, but story scores won't be updated until a refresh.
                // for best offline operation, we could try to duplicate that business logic locally
                dbHelper.clearClassifiersForFeed(feedId);
                classifier.feedId = feedId; 
                dbHelper.insertClassifier(classifier);
                impact |= UPDATE_INTEL;
                break;

            case RENAME_FEED:
                dbHelper.renameFeed(feedId, newFeedName);
                impact |= UPDATE_METADATA;
                break;

            default:
                // not all actions have these, which is fine
        }
        return impact;
    }

}

File: clients/android/NewsBlur/app/src/main/java/com/newsblur/util/ReadingAction.kt
package com.newsblur.util

import android.content.ContentValues
import android.database.Cursor
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.newsblur.database.BlurDatabaseHelper
import com.newsblur.database.DatabaseConstants
import com.newsblur.domain.Classifier
import com.newsblur.network.FeedApi
import com.newsblur.network.StoryApi
import com.newsblur.network.domain.CommentResponse
import com.newsblur.network.domain.NewsBlurResponse
import com.newsblur.network.domain.StoriesResponse
import com.newsblur.preference.PrefsRepo
import com.newsblur.service.NbSyncManager.UPDATE_INTEL
import com.newsblur.service.NbSyncManager.UPDATE_METADATA
import com.newsblur.service.NbSyncManager.UPDATE_SOCIAL
import com.newsblur.service.NbSyncManager.UPDATE_STORY
import com.newsblur.service.SyncServiceState
import com.newsblur.util.ReadingAction.Companion.toJson
import java.io.Serializable

sealed interface ReadingAction : Serializable {
    val time: Long
    val tried: Int

    data class MarkStoryRead(
            val storyHash: String,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class MarkStoryUnread(
            val storyHash: String,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class SaveStory(
            val storyHash: String,
            val highlights: List<String>,
            val userTags: List<String>,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class UnsaveStory(
            val storyHash: String,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class MarkFeedRead(
            val feedSet: FeedSet,
            val olderThan: Long? = null,
            val newerThan: Long? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class ShareStory(
            val storyHash: String? = null,
            val storyId: String? = null,
            val feedId: String? = null,
            val sourceUserId: String? = null,
            val commentReplyText: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class UnshareStory(
            val storyHash: String? = null,
            val storyId: String? = null,
            val feedId: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class LikeComment(
            val storyId: String? = null,
            val commentUserId: String? = null,
            val feedId: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class UnlikeComment(
            val storyId: String? = null,
            val commentUserId: String? = null,
            val feedId: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class ReplyToComment(
            val storyId: String? = null,
            val feedId: String? = null,
            val commentUserId: String? = null,
            val commentReplyText: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class EditReply(
            val storyId: String? = null,
            val feedId: String? = null,
            val commentUserId: String? = null,
            val replyId: String? = null,
            val commentReplyText: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class DeleteReply(
            val storyId: String? = null,
            val feedId: String? = null,
            val commentUserId: String? = null,
            val replyId: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class MuteFeeds(
            val activeFeedIds: Set<String>,
            val modifiedFeedIds: Set<String>,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class UnmuteFeeds(
            val activeFeedIds: Set<String>,
            val modifiedFeedIds: Set<String>,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class SetNotify(
            val feedId: String? = null,
            val notifyTypes: List<String> = emptyList(),
            val notifyFilter: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class InstaFetch(
            val feedId: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class UpdateIntel(
            val feedId: String? = null,
            val classifier: Classifier? = null,
            val feedSet: FeedSet? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    data class RenameFeed(
            val feedId: String? = null,
            val newFeedName: String? = null,
            override val time: Long = System.currentTimeMillis(),
            override val tried: Int = 0,
    ) : ReadingAction

    companion object {

        fun toJson(action: ReadingAction): String =
                gson.toJson(action, ReadingAction::class.java)

        fun fromJson(json: String): ReadingAction =
                gson.fromJson(json, ReadingAction::class.java)

        @JvmStatic
        fun toContentValues(action: ReadingAction): ContentValues = ContentValues().apply {
            put(DatabaseConstants.ACTION_TIME, action.time)
            put(DatabaseConstants.ACTION_TRIED, action.tried)
            put(DatabaseConstants.ACTION_PARAMS, toJson(action))
        }

        fun fromCursor(c: Cursor): ReadingAction {
            val time = c.getLong(c.getColumnIndexOrThrow(DatabaseConstants.ACTION_TIME))
            val tried = c.getInt(c.getColumnIndexOrThrow(DatabaseConstants.ACTION_TRIED))
            val params = c.getString(c.getColumnIndexOrThrow(DatabaseConstants.ACTION_PARAMS))
            val decoded = fromJson(params)

            return when (decoded) {
                is MarkStoryRead -> decoded.copy(time = time, tried = tried)
                is MarkStoryUnread -> decoded.copy(time = time, tried = tried)
                is SaveStory -> decoded.copy(time = time, tried = tried)
                is UnsaveStory -> decoded.copy(time = time, tried = tried)
                is MarkFeedRead -> decoded.copy(time = time, tried = tried)
                is ShareStory -> decoded.copy(time = time, tried = tried)
                is UnshareStory -> decoded.copy(time = time, tried = tried)
                is LikeComment -> decoded.copy(time = time, tried = tried)
                is UnlikeComment -> decoded.copy(time = time, tried = tried)
                is ReplyToComment -> decoded.copy(time = time, tried = tried)
                is EditReply -> decoded.copy(time = time, tried = tried)
                is DeleteReply -> decoded.copy(time = time, tried = tried)
                is MuteFeeds -> decoded.copy(time = time, tried = tried)
                is UnmuteFeeds -> decoded.copy(time = time, tried = tried)
                is SetNotify -> decoded.copy(time = time, tried = tried)
                is InstaFetch -> decoded.copy(time = time, tried = tried)
                is UpdateIntel -> decoded.copy(time = time, tried = tried)
                is RenameFeed -> decoded.copy(time = time, tried = tried)
            }
        }

        private val gson: Gson by lazy {
            val factory = RuntimeTypeAdapterFactory
                    .of(ReadingAction::class.java, "type")
                    .registerSubtype(MarkStoryRead::class.java, "MARK_READ_STORY")
                    .registerSubtype(MarkStoryUnread::class.java, "MARK_UNREAD_STORY")
                    .registerSubtype(MarkFeedRead::class.java, "MARK_READ_FEED")
                    .registerSubtype(SaveStory::class.java, "SAVE_STORY")
                    .registerSubtype(UnsaveStory::class.java, "UNSAVE_STORY")
                    .registerSubtype(ShareStory::class.java, "SHARE")
                    .registerSubtype(UnshareStory::class.java, "UNSHARE")
                    .registerSubtype(LikeComment::class.java, "LIKE_COMMENT")
                    .registerSubtype(UnlikeComment::class.java, "UNLIKE_COMMENT")
                    .registerSubtype(ReplyToComment::class.java, "REPLY")
                    .registerSubtype(EditReply::class.java, "EDIT_REPLY")
                    .registerSubtype(DeleteReply::class.java, "DELETE_REPLY")
                    .registerSubtype(MuteFeeds::class.java, "MUTE_FEEDS")
                    .registerSubtype(UnmuteFeeds::class.java, "UNMUTE_FEEDS")
                    .registerSubtype(SetNotify::class.java, "SET_NOTIFY")
                    .registerSubtype(InstaFetch::class.java, "INSTA_FETCH")
                    .registerSubtype(UpdateIntel::class.java, "UPDATE_INTEL")
                    .registerSubtype(RenameFeed::class.java, "RENAME_FEED")

            GsonBuilder()
                    .registerTypeAdapterFactory(factory)
                    .disableHtmlEscaping()
                    .create()
        }
    }
}

fun ReadingAction.toContentValues(): ContentValues = ContentValues().apply {
    put(DatabaseConstants.ACTION_TIME, time)
    put(DatabaseConstants.ACTION_TRIED, tried)
    put(DatabaseConstants.ACTION_PARAMS, toJson(this@toContentValues))
}

suspend fun ReadingAction.doRemote(
        syncServiceState: SyncServiceState,
        feedApi: FeedApi,
        storyApi: StoryApi,
        dbHelper: BlurDatabaseHelper,
        stateFilter: StateFilter
): NewsBlurResponse? {
    var result: NewsBlurResponse? = null
    var impact = 0

    suspend fun applyStoriesResponse(sr: StoriesResponse?) {
        if (sr != null) {
            result = sr
            if (sr.story != null) {
                dbHelper.updateStory(sr, stateFilter, true)
            } else {
                Log.w(this, "failed to refresh story data after action")
            }
            impact = impact or UPDATE_SOCIAL
        }
    }

    suspend fun applyCommentResponse(cr: CommentResponse?, storyId: String?) {
        if (cr != null) {
            result = cr
            if (cr.comment != null && storyId != null) {
                dbHelper.updateComment(cr, storyId)
            } else {
                Log.w(this, "failed to refresh comment data after action")
            }
            impact = impact or UPDATE_SOCIAL
        }
    }

    when (this) {
        // MARK_READ (story)
        is ReadingAction.MarkStoryRead ->
            result = storyApi.markStoryAsRead(storyHash)

        is ReadingAction.MarkStoryUnread ->
            result = storyApi.markStoryHashUnread(storyHash)

        is ReadingAction.SaveStory ->
            result = storyApi.markStoryAsStarred(storyHash, highlights, userTags)

        is ReadingAction.UnsaveStory ->
            result = storyApi.markStoryAsUnstarred(storyHash)

        // MARK_READ (feed range)
        is ReadingAction.MarkFeedRead ->
            result = feedApi.markFeedsAsRead(feedSet, olderThan, newerThan)

        is ReadingAction.ShareStory -> {
            val sr = storyApi.shareStory(storyId, feedId, commentReplyText, sourceUserId)
            applyStoriesResponse(sr)
        }

        is ReadingAction.UnshareStory -> {
            val sr = storyApi.unshareStory(storyId, feedId)
            applyStoriesResponse(sr)
        }

        is ReadingAction.LikeComment ->
            result = storyApi.favouriteComment(storyId, commentUserId, feedId)

        is ReadingAction.UnlikeComment ->
            result = storyApi.unFavouriteComment(storyId, commentUserId, feedId)

        is ReadingAction.ReplyToComment -> {
            val cr = storyApi.replyToComment(storyId, feedId, commentUserId, commentReplyText)
            applyCommentResponse(cr, storyId)
        }

        is ReadingAction.EditReply -> {
            val cr = storyApi.editReply(storyId, feedId, commentUserId, replyId, commentReplyText)
            applyCommentResponse(cr, storyId)
        }

        is ReadingAction.DeleteReply -> {
            val cr = storyApi.deleteReply(storyId, feedId, commentUserId, replyId)
            applyCommentResponse(cr, storyId)
        }

        is ReadingAction.MuteFeeds ->
            result = feedApi.saveFeedChooser(activeFeedIds)

        is ReadingAction.UnmuteFeeds ->
            result = feedApi.saveFeedChooser(activeFeedIds)

        is ReadingAction.SetNotify ->
            result = feedApi.updateFeedNotifications(feedId, notifyTypes, notifyFilter)

        is ReadingAction.InstaFetch -> {
            result = feedApi.instaFetch(feedId)
            feedId?.let {
                syncServiceState.addRecountCandidate(FeedSet.singleFeed(it))
                syncServiceState.flushRecounts()
            }
        }

        is ReadingAction.UpdateIntel -> {
            result = feedApi.updateFeedIntel(feedId, classifier)
            feedSet?.let { fs ->
                syncServiceState.resetFetchState(fs)
                syncServiceState.addRecountCandidate(fs)
            }
        }

        is ReadingAction.RenameFeed ->
            result = feedApi.renameFeed(feedId, newFeedName)
    }

    if (result != null && impact != 0) {
        result!!.impactCode = impact
    }

    return result
}

// TODO suspend
fun ReadingAction.doLocal(
        dbHelper: BlurDatabaseHelper,
        prefsRepo: PrefsRepo
): Int = doLocal(dbHelper, prefsRepo, isFollowup = false)

fun ReadingAction.doLocal(
        dbHelper: BlurDatabaseHelper,
        prefsRepo: PrefsRepo,
        isFollowup: Boolean
): Int {
    val userId = prefsRepo.getUserId()
    var impact = 0

    fun plus(flag: Int) {
        impact = impact or flag
    }

    when (this) {

        // MARK_READ (story)
        is ReadingAction.MarkStoryRead -> {
            dbHelper.setStoryReadState(storyHash, true)
            plus(UPDATE_METADATA)
            plus(UPDATE_STORY)
        }

        is ReadingAction.MarkStoryUnread -> {
            dbHelper.setStoryReadState(storyHash, false)
            plus(UPDATE_METADATA)
        }

        is ReadingAction.SaveStory -> {
            dbHelper.setStoryStarred(storyHash, highlights, userTags, true)
            plus(UPDATE_METADATA)
        }

        is ReadingAction.UnsaveStory -> {
            dbHelper.setStoryStarred(storyHash, emptyList(), emptyList(), false)
            plus(UPDATE_METADATA)
        }

        is ReadingAction.MarkFeedRead -> {
            dbHelper.markStoriesRead(feedSet, olderThan, newerThan)
            dbHelper.updateLocalFeedCounts(feedSet)
            plus(UPDATE_METADATA)
            plus(UPDATE_STORY)
        }

        is ReadingAction.ShareStory -> {
            if (!isFollowup) {
                // shares are only placeholders
                dbHelper.setStoryShared(storyHash, userId, true)
                dbHelper.insertCommentPlaceholder(storyId, userId, commentReplyText)
                plus(UPDATE_SOCIAL)
                plus(UPDATE_STORY)
            }
        }

        is ReadingAction.UnshareStory -> {
            dbHelper.setStoryShared(storyHash, userId, false)
            dbHelper.clearSelfComments(storyId, userId)
            plus(UPDATE_SOCIAL)
            plus(UPDATE_STORY)
        }

        is ReadingAction.LikeComment -> {
            dbHelper.setCommentLiked(storyId, commentUserId, userId, true)
            plus(UPDATE_SOCIAL)
        }

        is ReadingAction.UnlikeComment -> {
            dbHelper.setCommentLiked(storyId, commentUserId, userId, false)
            plus(UPDATE_SOCIAL)
        }

        is ReadingAction.ReplyToComment -> {
            if (!isFollowup) {
                dbHelper.insertReplyPlaceholder(storyId, userId, commentUserId, commentReplyText)
            }
        }

        is ReadingAction.EditReply -> {
            dbHelper.editReply(replyId, commentReplyText)
            plus(UPDATE_SOCIAL)
        }

        is ReadingAction.DeleteReply -> {
            dbHelper.deleteReply(replyId)
            plus(UPDATE_SOCIAL)
        }

        is ReadingAction.MuteFeeds -> {
            dbHelper.setFeedsActive(modifiedFeedIds, false)
            plus(UPDATE_METADATA)
        }

        is ReadingAction.UnmuteFeeds -> {
            dbHelper.setFeedsActive(modifiedFeedIds, true)
            plus(UPDATE_METADATA)
        }

        is ReadingAction.SetNotify -> {
            plus(UPDATE_METADATA)
        }

        // INSTA_FETCH: non-idempotent & purely graphical
        is ReadingAction.InstaFetch -> {
            if (!isFollowup && feedId != null) {
                dbHelper.setFeedFetchPending(feedId)
            }
        }

        is ReadingAction.UpdateIntel -> {
            dbHelper.clearClassifiersForFeed(feedId)
            classifier?.let { cls ->
                cls.feedId = feedId
                dbHelper.insertClassifier(cls)
            }
            plus(UPDATE_INTEL)
        }

        is ReadingAction.RenameFeed -> {
            dbHelper.renameFeed(feedId, newFeedName)
            plus(UPDATE_METADATA)
        }
    }

    return impact
}


